dim/src/external/tmdb/mod.rs

```diff
@@ -0,0 +1,475 @@
+// #![deny(warnings)]
+
+use std::collections::HashMap;
+use std::future::Future;
+use std::marker::PhantomData;
+
+use async_trait::async_trait;
+
+use displaydoc::Display;
+
+use tokio::sync::broadcast;
+
+use parking_lot::RwLock;
+
+use super::{Result as QueryResult, *};
+use core::result::Result;
+
+/// The User-Agent is generated from the package name "dim" and the version so "dim/0.x.y.z"
+pub const APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
+
+/// The base url used to access TMDB;
+pub const TMDB_BASE_URL: &str = "https://api.themoviedb.org/3";
+
+// -- TMDB API Data Models
+
+#[derive(Deserialize, Clone, Debug)]
+struct SearchResponse {
+    results: Vec<Option<TMDBMediaObject>>,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+struct TMDBMediaObject {
+    id: u64,
+    #[serde(rename(deserialize = "title", deserialize = "name"))]
+    title: String,
+    #[serde(rename(deserialize = "release_date", deserialize = "first_air_date"))]
+    release_date: Option<String>,
+    overview: Option<String>,
+    vote_average: Option<f64>,
+    poster_path: Option<String>,
+    backdrop_path: Option<String>,
+    genre_ids: Option<Vec<u64>>,
+    #[serde(skip_deserializing)]
+    genres: Vec<String>,
+    runtime: Option<u64>,
+}
+
+#[derive(Deserialize)]
+struct GenreList {
+    genres: Vec<Genre>,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct Genre {
+    id: u64,
+    name: String,
+}
+
+// -- TMDBClient
+
+#[derive(Debug, Display, Clone, thiserror::Error)]
+enum TMDBClientRequestError {
+    /// The body of a response was not value UTF-8.
+    InvalidUTF8Body,
+    /// the error comes from reqwest.
+    ReqwestError(#[from] Arc<reqwest::Error>),
+}
+
+impl TMDBClientRequestError {
+    fn reqwest(err: reqwest::Error) -> Self {
+        Self::ReqwestError(Arc::new(err))
+    }
+}
+
+/// Internal TMDB client type used for building and making requests.
+struct TMDBClient {
+    provider: TMDBMetadataProvider,
+}
+
+impl TMDBClient {
+    fn make_request<A, T>(
+        &self,
+        args: A,
+        path: String,
+    ) -> impl Future<Output = Result<String, TMDBClientRequestError>>
+    where
+        A: IntoIterator<Item = (T, T)>,
+        T: ToString,
+    {
+        let url = format!("{TMDB_BASE_URL}{path}");
+        let args: Vec<_> = args
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), v.to_string()))
+            .collect();
+
+        let request = self.provider.http_client.get(url).query(&args);
+
+        async move {
+            let response = request
+                .send()
+                .await
+                .map_err(TMDBClientRequestError::reqwest)?;
+
+            let body = response
```

Comment:
If tmdb returns a non-ok http code, we might want to deserialize the body returned as an error.